File: app/src/main/java/com/example/taxconnect/MilestonesActivity.java
package com.example.taxconnect;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.taxconnect.adapter.MilestoneAdapter;
import com.example.taxconnect.model.MilestoneModel;
import com.example.taxconnect.model.UserModel;
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.Timestamp;
import com.google.firebase.auth.FirebaseAuth;

import java.util.List;
import java.util.UUID;

public class MilestonesActivity extends AppCompatActivity {

    private RecyclerView rvMilestones;
    private MilestoneAdapter adapter;
    private ExtendedFloatingActionButton fabAddMilestone;
    private DataRepository repository;
    private String bookingId;
    private boolean isCa = false;

    private android.widget.TextView tvProgressPercent;
    private android.widget.ProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_milestones);

        bookingId = getIntent().getStringExtra("bookingId");
        if (bookingId == null) {
            finish();
            return;
        }

        repository = DataRepository.getInstance();
        rvMilestones = findViewById(R.id.rvMilestones);
        fabAddMilestone = findViewById(R.id.fabAddMilestone);
        tvProgressPercent = findViewById(R.id.tvProgressPercent);
        progressBar = findViewById(R.id.progressBar);

        findViewById(R.id.toolbar).setOnClickListener(v -> finish());

        checkUserRole();
    }

    private void checkUserRole() {
        String uid = FirebaseAuth.getInstance().getUid();
        if (uid == null) return;

        repository.fetchUser(uid, new DataRepository.DataCallback<UserModel>() {
            @Override
            public void onSuccess(UserModel user) {
                if (user != null) {
                    isCa = "CA".equals(user.getRole());
                    setupUI();
                }
            }

            @Override
            public void onError(String error) {
                // Handle error
            }
        });
    }

    private void setupUI() {
        if (isCa) {
            fabAddMilestone.setVisibility(View.VISIBLE);
            fabAddMilestone.setOnClickListener(v -> showAddMilestoneDialog());
        } else {
            fabAddMilestone.setVisibility(View.GONE);
        }

        adapter = new MilestoneAdapter(isCa, milestone -> {
            if (isCa) {
                toggleMilestoneStatus(milestone);
            }
        });
        rvMilestones.setLayoutManager(new LinearLayoutManager(this));
        rvMilestones.setAdapter(adapter);

        loadMilestones();
    }

    private void loadMilestones() {
        repository.getMilestones(bookingId, new DataRepository.DataCallback<List<MilestoneModel>>() {
            @Override
            public void onSuccess(List<MilestoneModel> data) {
                adapter.setMilestones(data);
                updateProgress(data);
            }

            @Override
            public void onError(String error) {
                Toast.makeText(MilestonesActivity.this, "Failed to load milestones", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void updateProgress(List<MilestoneModel> data) {
        if (data == null || data.isEmpty()) {
            progressBar.setProgress(0);
            tvProgressPercent.setText("0%");
            return;
        }
        
        int completed = 0;
        for (MilestoneModel m : data) {
            if ("COMPLETED".equals(m.getStatus())) {
                completed++;
            }
        }
        
        int progress = (int) ((completed / (float) data.size()) * 100);
        progressBar.setProgress(progress);
        tvProgressPercent.setText(progress + "%");
    }

    private void toggleMilestoneStatus(MilestoneModel milestone) {
        String currentStatus = milestone.getStatus();
        String newStatus;
        if ("PENDING".equals(currentStatus)) {
            newStatus = "IN_PROGRESS";
        } else if ("IN_PROGRESS".equals(currentStatus)) {
            newStatus = "COMPLETED";
        } else {
            newStatus = "PENDING"; // Cycle back
        }

        repository.updateMilestoneStatus(bookingId, milestone.getId(), newStatus, new DataRepository.DataCallback<Void>() {
            @Override
            public void onSuccess(Void data) {
                milestone.setStatus(newStatus);
                adapter.notifyDataSetChanged();
                // We need to reload or manually recalculate progress. 
                // Since adapter has the reference, we can re-iterate.
                // But simplest is to just call loadMilestones() again or calculate from adapter list.
                // Better: iterate adapter list.
                // However, we can just fetch list again for safety or assume local list is updated.
                // Let's assume local list is updated since we updated the object reference.
                // But we don't have easy access to the full list unless we expose it or keep a local copy in Activity.
                // Actually, loadMilestones passed 'data' to adapter. 'milestone' is a reference to an object in that list.
                // So if we have a reference to the list, we can recalc.
                // Let's just reload for consistency or keep a local list.
                loadMilestones(); 
            }

            @Override
            public void onError(String error) {
                Toast.makeText(MilestonesActivity.this, "Failed to update status", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void showAddMilestoneDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        View view = LayoutInflater.from(this).inflate(R.layout.dialog_add_milestone, null);
        builder.setView(view);
        AlertDialog dialog = builder.create();

        TextInputEditText etTitle = view.findViewById(R.id.etTitle);
        TextInputEditText etDescription = view.findViewById(R.id.etDescription);
        Button btnAdd = view.findViewById(R.id.btnAdd);
        Button btnCancel = view.findViewById(R.id.btnCancel);

        btnAdd.setOnClickListener(v -> {
            String title = etTitle.getText().toString().trim();
            String desc = etDescription.getText().toString().trim();

            if (title.isEmpty()) return;

            addMilestone(title, desc);
            dialog.dismiss();
        });

        btnCancel.setOnClickListener(v -> dialog.dismiss());
        dialog.show();
    }

    private void addMilestone(String title, String description) {
        MilestoneModel milestone = new MilestoneModel();
        milestone.setId(UUID.randomUUID().toString());
        milestone.setBookingId(bookingId);
        milestone.setTitle(title);
        milestone.setDescription(description);
        milestone.setStatus("PENDING");
        milestone.setTimestamp(Timestamp.now());

        repository.addMilestone(bookingId, milestone, new DataRepository.DataCallback<Void>() {
            @Override
            public void onSuccess(Void data) {
                Toast.makeText(MilestonesActivity.this, "Milestone Added", Toast.LENGTH_SHORT).show();
                loadMilestones();
            }

            @Override
            public void onError(String error) {
                Toast.makeText(MilestonesActivity.this, "Failed to add milestone", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: tax connect/app/src/main/java/com/example/taxconnect/features/milestones/MilestonesActivity.kt
package com.example.taxconnect.features.milestones

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.taxconnect.R
import com.example.taxconnect.data.repositories.DataRepository
import com.example.taxconnect.data.repositories.DataRepository.DataCallback
import com.example.taxconnect.features.milestones.MilestoneAdapter
import com.example.taxconnect.databinding.ActivityMilestonesBinding
import com.example.taxconnect.data.models.MilestoneModel
import com.example.taxconnect.data.models.UserModel
import com.example.taxconnect.core.base.BaseActivity
import com.google.android.material.textfield.TextInputEditText
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import java.util.UUID

class MilestonesActivity : BaseActivity<ActivityMilestonesBinding>() {
    override val bindingInflater: (LayoutInflater) -> ActivityMilestonesBinding = ActivityMilestonesBinding::inflate
    
    private lateinit var adapter: MilestoneAdapter
    private lateinit var repository: DataRepository
    private var bookingId: String? = null
    private var isCa = false

    override fun initViews() {
        bookingId = intent.getStringExtra("bookingId")
        if (bookingId == null) {
            finish()
            return
        }

        repository = DataRepository.getInstance()

        checkUserRole()
    }

    override fun setupListeners() {
        binding.toolbar.setOnClickListener { finish() }
    }

    override fun observeViewModel() {
        // No ViewModel to observe yet
    }

    private fun checkUserRole() {
        val uid = FirebaseAuth.getInstance().uid ?: return

        repository.fetchUser(uid, object : DataCallback<UserModel> {
            override fun onSuccess(data: UserModel?) {
                if (data != null) {
                    isCa = "CA" == data.role
                    setupUI()
                }
            }

            override fun onError(error: String?) {
                // Handle error
            }
        })
    }

    private fun setupUI() {
        if (isCa) {
            binding.fabAddMilestone.visibility = View.VISIBLE
            binding.fabAddMilestone.setOnClickListener { showAddMilestoneDialog() }
        } else {
            binding.fabAddMilestone.visibility = View.GONE
        }

        adapter = MilestoneAdapter(isCa) { milestone ->
            if (isCa) {
                toggleMilestoneStatus(milestone)
            }
        }
        binding.rvMilestones.layoutManager = LinearLayoutManager(this)
        binding.rvMilestones.adapter = adapter

        loadMilestones()
    }

    private fun loadMilestones() {
        val bid = bookingId ?: return
        repository.getMilestones(bid, object : DataCallback<List<MilestoneModel>> {
            override fun onSuccess(data: List<MilestoneModel>?) {
                if (data != null) {
                    adapter.setMilestones(data)
                    updateProgress(data)
                }
            }

            override fun onError(error: String?) {
                showToast(getString(R.string.failed_to_load_milestones))
            }
        })
    }

    private fun updateProgress(data: List<MilestoneModel>?) {
        if (data.isNullOrEmpty()) {
            binding.progressBar.progress = 0
            binding.tvProgressPercent.text = "0%"
            return
        }

        var completed = 0
        for (m in data) {
            if ("COMPLETED" == m.status) {
                completed++
            }
        }

        val progress = (completed / data.size.toFloat() * 100).toInt()
        binding.progressBar.progress = progress
        binding.tvProgressPercent.text = "$progress%"
    }

    private fun toggleMilestoneStatus(milestone: MilestoneModel) {
        val bid = bookingId ?: return
        val mid = milestone.id ?: return
        val currentStatus = milestone.status
        val newStatus = when (currentStatus) {
            "PENDING" -> "IN_PROGRESS"
            "IN_PROGRESS" -> "COMPLETED"
            else -> "PENDING"
        }

        repository.updateMilestoneStatus(
            bid,
            mid,
            newStatus,
            object : DataCallback<Void?> {
                override fun onSuccess(data: Void?) {
                    milestone.status = newStatus
                    adapter.notifyDataSetChanged()
                    loadMilestones()
                }

                override fun onError(error: String?) {
                    showToast(getString(R.string.failed_to_update_status))
                }
            })
    }

    private fun showAddMilestoneDialog() {
        val builder = AlertDialog.Builder(this)
        val view = LayoutInflater.from(this).inflate(R.layout.dialog_add_milestone, null)
        builder.setView(view)
        val dialog = builder.create()

        val etTitle = view.findViewById<TextInputEditText>(R.id.etTitle)
        val etDescription = view.findViewById<TextInputEditText>(R.id.etDescription)
        val btnAdd = view.findViewById<Button>(R.id.btnAdd)
        val btnCancel = view.findViewById<Button>(R.id.btnCancel)

        btnAdd.setOnClickListener {
            val title = etTitle.text.toString().trim()
            val desc = etDescription.text.toString().trim()

            if (title.isEmpty()) {
                etTitle.error = getString(R.string.milestone_title_required)
                etTitle.requestFocus()
                return@setOnClickListener
            }

            addMilestone(title, desc)
            dialog.dismiss()
        }

        btnCancel.setOnClickListener { dialog.dismiss() }
        dialog.show()
    }

    private fun addMilestone(title: String, description: String) {
        val bid = bookingId ?: return
        val milestone = MilestoneModel()
        milestone.id = UUID.randomUUID().toString()
        milestone.bookingId = bid
        milestone.title = title
        milestone.description = description
        milestone.status = "PENDING"
        milestone.timestamp = Timestamp.now()

        repository.addMilestone(bid, milestone, object : DataCallback<Void?> {
            override fun onSuccess(data: Void?) {
                showToast("Milestone Added")
                loadMilestones()
            }

            override fun onError(error: String?) {
                showToast("Failed to add milestone")
            }
        })
    }
}
